--- target/linux/mvebu/base-files/etc/uci-defaults/02_network	2016-04-17 11:29:39.000000000 +0800
+++ target/linux/mvebu/base-files/etc/uci-defaults/02_network	2016-05-09 22:08:26.071472307 +0800
@@ -17,6 +17,9 @@
 case "$board" in
 armada-385-linksys-caiman)
 	ucidef_set_interfaces_lan_wan "eth1" "eth0"
+	ucidef_add_switch "switch0" "1" "1"
+	ucidef_add_switch_vlan "switch0" "1" "0 1 2 3 6"
+	ucidef_add_switch_vlan "switch0" "2" "4 5"
 	;;
 armada-385-linksys-cobra)
 	ucidef_set_interfaces_lan_wan "eth1" "eth0"
